src/sentry/static/sentry/app/components/gridEditable/grid.tsx

```diff
@@ -0,0 +1,304 @@
+import React from 'react';
+
+import {t} from 'app/locale';
+import {openModal} from 'app/actionCreators/modal';
+
+import Alert from 'app/components/alert';
+import EmptyStateWarning from 'app/components/emptyStateWarning';
+import InlineSvg from 'app/components/inlineSvg';
+import LoadingContainer from 'app/components/loading/loadingContainer';
+import ToolTip from 'app/components/tooltip';
+
+import {GridColumn, GridColumnOrder, GridColumnSortBy} from './gridTypes';
+import GridHeadCell from './gridHeadCell';
+import GridModalEditColumn from './gridModalEditColumn';
+import {
+  GridPanel,
+  GridPanelBody,
+  Grid,
+  GridRow,
+  GridHead,
+  GridBody,
+  GridBodyCell,
+  GridBodyCellSpan,
+  GridBodyCellLoading,
+  GridEditGroup,
+  GridEditGroupButton,
+} from './styles';
+
+export type GridEditableProps<DataRow, ColumnKey extends keyof DataRow> = {
+  isEditable?: boolean;
+  isLoading?: boolean;
+  error?: React.ReactNode | null;
+
+  /**
+   * GridEditable (mostly) do not maintain any internal state and relies on the
+   * parent component to tell it how/what to render and will mutate the view
+   * based on this 3 main props.
+   *
+   * - `columnOrder` determines the columns to show, from left to right
+   * - `columnSortBy` is not used at the moment, however it might be better to
+   *   move sorting into Grid for performance
+   */
+  columnOrder: GridColumnOrder<ColumnKey>[];
+  columnSortBy: GridColumnSortBy<ColumnKey>[];
+  data: DataRow[];
+
+  /**
+   * GridEditable allows the parent component to determine how to display the
+   * data within it. Note that this is optional.
+   */
+  grid: {
+    renderHeaderCell?: (column: GridColumnOrder<ColumnKey>) => React.ReactNode;
+    renderBodyCell?: (
+      column: GridColumnOrder<ColumnKey>,
+      dataRow: DataRow
+    ) => React.ReactNode;
+  };
+
+  /**
+   * As GridEditable is unopinionated about the structure of GridColumn,
+   * ModalEditColumn relies on the parent component to provide the form layout
+   * and logic to create/update the columns
+   */
+  modalEditColumn: {
+    renderBodyWithForm: (
+      indexOfColumnOrder?: number,
+      column?: GridColumn<ColumnKey>
+    ) => React.ReactNode;
+    renderFooter: () => React.ReactNode;
+  };
+
+  /**
+   * As there is no internal state being maintained, the parent component will
+   * have to provide functions to move/delete the columns
+   */
+  actions: {
+    moveColumn: (fromIndex: number, toIndex: number) => void;
+    deleteColumn: (index: number) => void;
+  };
+};
+export type GridEditableState = {
+  isEditing: boolean;
+  numColumn: number;
+};
+
+class GridEditable<DataRow extends Object> extends React.Component<
```

Comment:
Should the component be named `Grid` to match the filename?

Reply:
I did it the other way and renamed the file to `index.tsx` per the convention mentioned below.